apps/aecore/lib/aecore/contract/vm_chain.ex

```diff
@@ -0,0 +1,134 @@
+        store = Contract.store(contract)
+
+        Enum.reduce(store, %{}, fn {key, value}, acc ->
+          <<key_integer::256>> = key
+          <<value_integer::256>> = value
+          Map.put(acc, key_integer, value_integer)
+        end)
+
+      :none ->
+        %{}
+    end
+  end
+
+  @spec set_store(ChainApi.chain_state(), ChainApi.store()) :: ChainApi.chain_state()
+  def set_store(store, %{pubkey: pubkey, chain_state: chain_state} = state) do
+    contract_tree = chain_state.contracts
+    contract = ContractStateTree.get_contract(contract_tree, pubkey)
+
+    new_contract = %Contract{contract | store: store}
+    updated_contracts_tree = ContractStateTree.enter_contract(contract_tree, new_contract)
+
+    %{state | chain_state: %{chain_state | contracts: updated_contracts_tree}}
+  end
+
+  @spec call_contract(
+          Keys.pubkey(),
+          non_neg_integer(),
+          non_neg_integer(),
+          binary(),
+          [non_neg_integer()],
+          ChainApi.chain_state()
+        ) :: {:ok, ChainApi.call_result(), ChainApi.chain_state()} | {:error, String.t()}
+  def call_contract(
+        target,
+        gas,
+        value,
+        call_data,
+        call_stack,
+        %{pubkey: contract_key, chain_state: chain_state}
+      ) do
+    contract_tree = chain_state.contracts
+
+    case ContractStateTree.get_contract(contract_tree, target) do
+      %Contract{} = contract ->
+        account_tree = chain_state.accounts
+        nonce = Account.nonce(account_tree, contract_key) + 1
+        vm_version = contract.vm_version
+
+        payload = %{
+          caller: contract_key,
+          contract: target,
+          vm_version: vm_version,
+          amount: value,
+          gas: gas,
+          gas_price: 0,
+          call_data: call_data,
+          call_stack: call_stack
+        }
+
+        height = Chain.top_height() + 1
+        tx = DataTx.init(ContractCallTx, payload, contract_key, 0, nonce)
+        do_call_contract(chain_state, height, tx)
+
+      :none ->
+        {:error, "#{__MODULE__}: No such contract: #{inspect(target)}"}
+    end
+  end
+
+  defp do_call_contract(
+         chain_state,
+         height,
+         %DataTx{
+           nonce: nonce,
+           payload: %{caller: contract_key, contract: target}
+         } = tx
+       ) do
+    with :ok <- DataTx.preprocess_check(chain_state, height, tx),
+         {:ok, new_chain_state} <- DataTx.process_chainstate(chain_state, height, tx) do
+      call_id = Call.id(contract_key, nonce, target)
+      call_tree_key = CallStateTree.construct_call_tree_id(contract_key, call_id)
+      call = CallStateTree.get_call(new_chain_state.calls, call_tree_key)
+      gas_used = call.gas_used
+
+      result =
+        case call.return_type do
+          return_type when return_type in [:ok, :revert] ->
+            %{result: call.return_value, gas_spent: gas_used}
+
+          :error ->
+            %{result: :out_of_gas, gas_spent: gas_used}
```

Comment:
If `:error` means `:out_of_gas` why not call it that from the beginning? (If it doesn't then probably it shouldn't be called like that here.)